packages/eslint-plugin/src/rules/prefer-at.ts

```diff
@@ -0,0 +1,177 @@
+    type: 'suggestion',
+    fixable: 'code',
+    docs: {
+      description:
+        'Enforce the use of `array.at(-1)` instead of `array[array.length - 1]`',
+      recommended: false,
+      requiresTypeChecking: true,
+    },
+    messages: {
+      preferAt:
+        'Expected a `{{name}}.at(-1)` instead of `{{name}}[{{name}}.length - 1]`.',
+    },
+    schema: [
+      {
+        oneOf: [
+          {
+            type: 'object',
+            properties: {
+              ignoreFunctions: {
+                type: 'boolean',
+              },
+            },
+            additionalProperties: false,
+          },
+        ],
+      },
+    ],
+  },
+  defaultOptions: [
+    {
+      ignoreFunctions: false,
+    },
+  ],
+  create(context, [options]) {
+    const parserServices = util.getParserServices(context);
+    const checker = parserServices.program.getTypeChecker();
+    const sourceCode = context.getSourceCode();
+
+    function getName(node: TSESTree.Node): string | undefined {
+      switch (node.type) {
+        case AST_NODE_TYPES.Identifier:
+          return node.name;
+        case AST_NODE_TYPES.MemberExpression:
+          return getName(node.property);
+        default:
+          return undefined;
+      }
+    }
+
+    function getFullName(node: TSESTree.Node): string {
+      return sourceCode.text.slice(node.range[0], node.range[1]);
+    }
+
+    function hasCallExpression(node: TSESTree.MemberExpression): boolean {
+      return (
+        node.object.type === AST_NODE_TYPES.CallExpression ||
+        (node.object.type === AST_NODE_TYPES.MemberExpression &&
+          hasCallExpression(node.object))
+      );
+    }
+
+    function getTypeAtLocation(node: TSESTree.Node): ts.Type {
+      return checker.getTypeAtLocation(
+        parserServices.esTreeNodeToTSNodeMap.get(node),
+      );
+    }
+
+    type SupportedObject = (type: ts.Type) => boolean;
+
+    function checkObjectName(name: string): SupportedObject {
+      return type => type.getSymbol()?.name === name;
+    }
+
+    function checkObjectType(flags: ts.TypeFlags): SupportedObject {
+      return type => type.getFlags() === flags;
+    }
+
+    const supporterObjects: Array<SupportedObject> = [
+      checkObjectName('Array'),
+      checkObjectName('Int8Array'),
+      checkObjectName('Uint8Array'),
+      checkObjectName('Uint8ClampedArray'),
+      checkObjectName('Int16Array'),
+      checkObjectName('Uint16Array'),
+      checkObjectName('Int32Array'),
+      checkObjectName('Float32Array'),
+      checkObjectName('Uint32Array'),
+      checkObjectName('Float64Array'),
+      checkObjectName('BigInt64Array'),
+      checkObjectName('BigUint64Array'),
+      // eslint-disable-next-line @typescript-eslint/internal/prefer-ast-types-enum
+      checkObjectName('String'),
+      checkObjectType(ts.TypeFlags.String),
+    ];
+
+    function isSupportedObject(type: ts.Type): boolean {
+      return supporterObjects.some(check => check(type));
+    }
```

Comment:
[Refactor] This `supporterObjects` schema took me a while to read through. If I understand it right, it roughly:

1. Checks if the type's symbol is one of the supported names
2. Checks if the object has `ts.TypeFlags.String` in its flags

If that's roughly right, two change suggestions: 

* I think it'd be a lot simpler to move those names into a `Set<string | undefined>` and use `.has`
* We normally use `tsutils.isTypeFlagSet` for checking type flags (as you've already done elsewhere in this file) - why not use it here?

```js
supportedObjectNames.has(type.getSymbol()?.name) || tsutils.isTypeFlagSet(type, ts.TypeFlags.String);
```

Thoughts?